homeassistant/helpers/network.py

```diff
@@ -5,8 +5,128 @@
+        if (not require_ssl or internal_url.scheme == "https") and (
+            not require_standard_port or internal_url.is_default_port()
+        ):
+            return (
+                str(internal_url)
+                if not internal_url.is_default_port()
+                else str(internal_url.with_port(None))
+            )
+
+    # Return IP if local is allowed and it matches citeria requested.
+    # This cannot be used if SSL is required, as IP based will always result
+    # in a invalid SSL certificate.
+    if (
+        allow_local
+        and not require_ssl
+        and hass.config.api is not None
+        and not hass.config.api.use_ssl
+    ):
+        ip_url = yarl.URL.build(
+            scheme="http", host=get_local_ip(), port=hass.config.api.port
+        )
+        if not is_loopback(ip_address(ip_url.host)) and (
+            not require_standard_port or ip_url.is_default_port()
+        ):
+            return (
+                str(ip_url)
+                if not ip_url.is_default_port()
+                else str(ip_url.with_port(None))
+            )
+
+    # Return user set external URL if it matches the requested citeria
+    if hass.config.external_url:
+        external_url = yarl.URL(hass.config.external_url.rstrip("/"))
+        if (not require_standard_port or external_url.is_default_port()) and (
+            not require_ssl
+            or (
+                external_url.scheme == "https"
+                and not is_ip_address(str(external_url.host))
+            )
+        ):
+            return (
+                str(external_url)
+                if not external_url.is_default_port()
+                else str(external_url.with_port(None))
+            )
+
+    # Fall back to the good old, deprecated `base_url`, with sanity
+    # This if statement can be removed after `base_url` is removed.
+    if hass.config.api is not None and hass.config.api.base_url:
+        base_url = yarl.URL(hass.config.api.base_url.rstrip("/"))
+        if (
+            (
+                not is_ip_address(str(base_url.host))
+                or (
+                    is_ip_address(str(base_url.host))
+                    and not is_loopback(ip_address(base_url.host))
+                )
+            )
+            and (
+                not require_ssl
+                or (
+                    base_url.scheme == "https" and not is_ip_address(str(base_url.host))
+                )
+            )
+            and (not require_standard_port or base_url.is_default_port())
+            and (
+                allow_local
+                or not is_ip_address(str(base_url.host))
+                or (
+                    is_ip_address(str(base_url.host))
+                    and not is_local(ip_address(base_url.host))
+                )
+            )
+        ):
+            return (
+                str(base_url)
+                if not base_url.is_default_port()
+                else str(base_url.with_port(None))
+            )
```

Comment:
can we make this a helper, `normalize_url` or something ?

Reply:
Done!